### ISSUE TITLE
gdalwarp destroying geotiff by removing geoTags or filling in paddings

### ISSUE THREAD

**Issue by VinzenzBauer** [OSGeo/gdal#641]
## Expected behavior and actual behavior.

I expect gdalwarp to not discard any GeoTiff-Tags, as only complete they form a valid geotiff.
And if gdalwarp sees the need to insert paddings (0,0,0,0), they should not be counted as keys, for different reason like there is no geokey == 0.

## Steps to reproduce the problem.

take a valid geotiff
1. gdalwarp -t_srs '+proj=utm +zone=11 +datum=WGS84' valid_geotiff.tif invalid_warp_geotiff.tif
in "invalid_warp_geotiff.tif" the "GeoDoubleParamsTag" (34736) will be removed !?!?
2. gdalwarp valid_geotiff.tif invalid_padding_geotiff.tif -t_srs "+proj=longlat +ellps=WGS84"
in "invalid_warp_geotiff.tif" the last key from GeoTiffDirectory:count will be a 0, 0, 0, 0 padding and be counted as a GeoKey, which is against the Specification of GeoTiff in generell. Also it will not match to the count of the first GeoKeyHeader:count which counts the following GeoKeys.
Example: The GeoTiff has 5 GeoKeys and this value will be correct in 1st GeoKeyHeader:count (value==5).
The GeoKeyDirectory:Count however should have the value (5+1)*4
The 1 beeing the header, the 4 beeing the 2Byte Decimals per Key, the 5 the KeyCount obviously.
But the value will be (5+2)*4, because a padding was included and that padding was falsly counted as a key, which is a wrong thing to do. There is no GeoKey 0. And they also have to be in increasing order, always!
Please review the GeoTiff Spec if you think otherwise.

## Operating system

Distributor ID: Debian
Description:    Debian GNU/Linux 9.4 (stretch)
Release:        9.4
Codename:       stretch

## GDAL version and provenance

GDAL 2.3.0, released 2018/05/04

**Comment by VinzenzBauer** [comment]
sure:
valid original: https://www.dropbox.com/s/vs9r51dmadcxe5r/valid_minimal.tif?dl=0
invalid padding: https://www.dropbox.com/s/rtje1lgdptsd22l/invalid_padding.tif?dl=0
invalid missing asciiparamstag: https://www.dropbox.com/s/7lknkay42jr7ubh/invalid_missing_asciiparamstag.tif?dl=0
i can send more details from okteta, if you need.

**Comment by jratike80** [comment]
Thanks. I am not a developer and don't need the annexes myself but I asked you to add them for making the ticket better.

**Comment by rouault** [comment]
Confirmed on the samples provided and I can reproduce with gdalwarp, at least since GDAL 1.11

**Comment by rouault** [comment]
I'm working on this. This is indeed a bug in libgeotiff. Will report the fix there too

**Comment by VinzenzBauer** [comment]
Thanks to roulault the padding-Problem (no. 2) seems to be fixed. Should i open a separate ticket for the removal of geodoubleparams (no. 1) ?

**Comment by rouault** [comment]
> Should i open a separate ticket for the removal of geodoubleparams (no. 1) ?

I'm not completely sure this is an issue. "+proj=utm +zone=11 +datum=WGS84" is completely captured in the output file with the ProjectedCSTypeGeoKey (Short,1): PCS_WGS84_UTM_zone_11N . The re-definitions of the WGS84 ellipsoid parameters in the 34736 would be redundant. 
If you look at the GeoTIFF spec example at http://geotiff.maptools.org/spec/geotiff3.html#3.1.1. , there is a pretty similar case and only ProjectedCSTypeGeoKey is defined, without redefinitions of GeogSemiMajorAxisGeoKey and GeogInvFlatteningGeoKey
And the "Cookbook" http://geotiff.maptools.org/spec/geotiff2.7.html#2.7.3 mentions for a projected coordinate system "If coded, it will not be necessary to specify the Projection datum, etc for this case, since all of those parameters are determined by the ProjectedCSTypeGeoKey code"

**Comment by VinzenzBauer** [comment]
> The re-definitions of the WGS84 ellipsoid parameters in the 34736 would be redundant.
You are right. The whole Point 1 was because of my understanding that tags doubleparams and asciiparams should always be there, even if empty. But this is not the case. Thanks for the clarification.
